### PR TITLE
server: Assume requested SR was successfully set

### LINKED ISSUE #768: s.options.sampleRate_ setter requires boot then reboot to take effect
Changing the sample rate using s.options.sampleRate_(sr) doesn't change the sample rate upon reboot if you're booting to a new sample rate from what was formerly running.  The user must boot (server boots to the old SR), quit, then reboot for the new SR to take effect.
Observe...
(assuming you were previously booted to 44.1 or 48...)
s.quit;
s.options.sampleRate_(96000);
s.boot; // still boots to the old sample rate
s.quit;
s.boot; // server now boots to the specified SR

Thanks in advance!

### PR DESCRIPTION
**Pull request by scztt**
This fixes #768 
Setting sample rate in CoreAudio is actually an asynchronous operation - you request a sample rate change, and some time in the future it (probably) happens and you are called back about it. This is a problem for us because (a) SC doesn't support the sample rate changes once the server is started, and (b) our startup routine is totally synchronous anyway. Prior to this change, when we requested a custom sample rate, we would do the SetProperty and then shortly after ask CoreAudio for the device sample rate. If the operation had not been completed yet, we would get the OLD rate - the device would set it's SR to the requested shortly after, and we would effectively be running the server at the wrong rate. Now, we *assume* that a successful SetProperty has changed the sample rate, and just use the rate we set instead of requesting it from CoreAudio. We can still be incorrect if the SR doesn't actually change, but at least we're correct MOST of the time. To increase the chances of success, we query CoreAudio for valid sample rates, and set our requested rate based on those.

A proper fix for this would be to have a properly asynchronous driver startup procedure, where we request all the properties we want, and then wait for callbacks indicating the changes have "taken" before continuing. Note that requesting a buffer size still has the same problem - the fix should be identical - this should be fixed in a different CL.

### CODE REVIEW

**Comment by scztt** [comment]
Can mac users carefully review and test this? This is the best change I could come up with, short is a large refactor of SC_CoreAudio - which is way out of scope at the moment. Specifically, test that launching the server w/ custom sample rates actually changes the device rate, and that playback is correct when launched (try a simple `{ SinOsc.ar(440) }.play;`). Finally check that unsupported sample rates fall back to a reasonable value (i.e. requesting 44099 should settle for setting to 44100), and that you can set sample rates for cases where your input and output devices support different sets of sample rates.

As noted, the same fix is required for buffer sizes, I'll do that soon.

File: server/scsynth/SC_CoreAudio.cpp [hunk 1]
```diff
@@ -746,29 +790,102 @@ bool SC_CoreAudioDriver::DriverSetup(int* outNumSamplesPerCallback, double* outS
...
+				count = sizeof(Float64);
+
+				propertyAddress.mSelector = kAudioDevicePropertyNominalSampleRate;
+				propertyAddress.mScope = kAudioDevicePropertyScopeInput;
+
+				err = AudioObjectSetPropertyData(mInputDevice, &propertyAddress, 0, NULL, count, &sampleRate);
+				if (err != kAudioHardwareNoError) {
+					scprintf("set kAudioDevicePropertyNominalSampleRate error %4.4s\n", (char*)&err);
+				} else {
+					mExplicitSampleRate = sampleRate;
```
Comment (scztt) [inline server/scsynth/SC_CoreAudio.cpp]: This assign needs to be moved to outside the `if (UseSeparateI())` scope.

**Comment by telephon** [comment]
I'm just testing this branch, it seems to behave a little erratic on os x. What should be tested specifically? 

I can start the server when ````s.options.sampleRate```` matches the current sample rate of the device. When I change the device rate, ````s.options.sampleRate```` (of course) doesn't change, but within 10 seconds or so,  ````s.actualSampleRate```` adjusts to the rate. But a running ````SinOsc```` will change frequency (at runtime) when I do this.

I hope this is helpful.

**Comment by scztt** [comment]
These test cases that should work:

1. Sample rate should be hardware sample rate.
```
s.options.sampleRate = nil;
s.reboot;
s.waitForBoot { {SinOsc.ar(440)}.play };
```

2. Sample rate should be 44100, tone should play correctly
```
s.options.sampleRate = 44100;
s.reboot;
s.waitForBoot { {SinOsc.ar(440)}.play };
```

3. Sample rate should be 48000, tone should play correctly
```
s.options.sampleRate = 48000;
s.reboot;
s.waitForBoot { {SinOsc.ar(440)}.play };
```

4. Sample rate should be 48000, warning should be shown, tone should play correctly
```
s.options.sampleRate = 48000 - 1;
s.reboot;
s.waitForBoot { {SinOsc.ar(440)}.play };
```

In all cases, after boot, s.actualSampleRate should converge to the correct sample almost immediately.
Before this fix, case 1 worked and cases 2-4 only worked if options.sampleRate matched the current hardware rate (i.e. no change was required).

**Comment by telephon** [comment]
@scztt all your tests work fine here, so this looks good.

The fact that changing the sample rate externally while the server is running causes a detune of the running synths is probably an acceptable oddity. If it is possible to post a warning, this might be good.

Btw. when the sample rate is changed externally, ````{ SampleRate.ir.poll }.play```` remains the same, but ````{SinOsc.ar(440)}.play```` changes (actual) frequency.

**Comment by telephon** [comment]
>Btw. when the sample rate is changed externally, { SampleRate.ir.poll }.play remains the same, but {SinOsc.ar(440)}.play changes (actual) frequency.

This is the same as it was before, actually, so should be fine.

**Comment by scztt** [comment]
Yeah - I looked some for my own edification and it's pretty clear that SuperCollider will not ever support changing the sample rate without a server restart, as it would require re-initialization of basically everything (which is, in essence, just restarting the server anyway). 

I'm inclined to say we should actually silence the server if the SR changes, so it's totally clear that something is wrong - rather than outputting bad audio that, frankly, could easily be interpreted as okay depending on the kinds of sounds you're making :)

**Comment by telephon** [comment]
yes, definitely ... :)

**Pull request merged**